python/pfs/pipe2d/generateReductionSpec.py

```diff
@@ -765,3 +766,192 @@ def nameYamlMapping(name: str, mapping: Dict[str, Any]) -> Dict[str, Any]:
     newMapping = {"name": name}
     newMapping.update((key, value) for key, value in mapping.items() if key != "name")
     return newMapping
+
+
+def mergeCalibBlocks(blocks: Iterable[Any]) -> List[Any]:
+    """Find mergeable calibBlocks in ``blocks`` and merge them.
+    Two calibBlocks are "mergeable"
+    if they are equal up to strings "arm=..." in them.
+
+    Parameters
+    ----------
+    blocks : `Iterable[Any]`
+        List of calibBlocks.
+
+    Returns
+    -------
+    newBlocks : `List[Any]`
+        Shallow copy of ``blocks``,
+        except that mergeable elements have been merged.
+    """
+    blocks = list(blocks)
+
+    newBlocks = []
+    while blocks:
+        mergeables = [0]
+        for i in range(1, len(blocks)):
+            if _mergeCalibBlocks_isMergeable(blocks[0], blocks[i]):
+                mergeables.append(i)
+        newBlocks.append(_mergeCalibBlocks_merge(blocks[i] for i in mergeables))
+        for i in reversed(mergeables):
+            del blocks[i]
+
+    return newBlocks
+
+
+def _mergeCalibBlocks_isMergeable(object1: Any, object2: Any) -> bool:
+    """Compare two calibBlocks and return True if they are mergeable,
+    which means they are equal up to strings "arm=...".
+    ("name" key is also ignored in comparison.)
+
+    Parameters
+    ----------
+    object1 : `Any`
+        Any constituent part of a calibBlock
+        (including the complete calibBlock itself).
+    object2 : `Any`
+        Any constituent part of a calibBlock
+        (including the complete calibBlock itself).
+
+    Returns
+    -------
+    isMergeable : `bool`
+        True if ``object1`` and ``object2`` are mergeable
+    """
+    if isinstance(object1, list):
+        if not isinstance(object2, list):
+            return False
+        if len(object1) != len(object2):
+            return False
+        return all(
+            _mergeCalibBlocks_isMergeable(elem1, elem2)
+            for elem1, elem2 in zip(object1, object2)
+        )
+
+    if isinstance(object1, dict):
+        if not isinstance(object2, dict):
+            return False
+        if set(object1.keys()) != set(object2.keys()):
+            return False
+        return all(
+            _mergeCalibBlocks_isMergeable(object1[key], object2[key])
+            for key in object1 if key != "name"
+        )
+
+    if isinstance(object1, str):
+        if not isinstance(object2, str):
+            return False
+        if object1.startswith("arm=") and object2.startswith("arm="):
+            return True
+        return object1 == object2
+
+    return object1 == object2
+
+
+def _mergeCalibBlocks_merge(objects: Iterable[Any]) -> Any:
+    """Merge calibBlocks.
+
+    Parameters
+    ----------
+    objects : `Iterable[Any]`
+        A list, each element being any constituent part of a calibBlock.
+        (including a complete calibBlock itself).
+        All elements must be mergeable to each other.
+        Otherwise ValueError is raised.
```

Comment:
I suggest putting this under a `Raises` heading.

Reply:
I deleted the phrase. In the course of writing this code, there was a time when this `ValueError` was actually caught by a caller. At that time ValueError had to be a part of API. But it no longer has to be in the current code.